---
title: "R Notebook"
output: html_notebook
---

# libraries
```{r}
library(tidyverse)
library(reshape2)
library(cowplot)
theme_set(theme_cowplot())
library(rtracklayer)
library(data.table)
```

# prepare data
```{r}
library(tidyverse)

dfpeaks <- read.csv("~/Documents/projects/ATAC_seq/data/ahyp_v3/peak_calling/ahypv3_w_sampleinfo_nowgs_2minov.csv", sep = ",")

library(GenomicRanges)
GRref <- makeGRangesFromDataFrame(dfpeaks)

detach("package:tidyverse")
detach("package:tidyr")
detach("package:lubridate")
detach("package:purrr")


# libraries for analysis
library(ChIPpeakAnno)
library(GenomicRanges)
library(GenomicFeatures)

GRref <- reduce(GRref)

library(tidyverse)


```


```{r   peak depth & methylation up and downstream of TSS}

setwd("~/Documents/projects/ATAC_seq/data/ref_genomes")
gff_ref <- import.gff("Ahypochondriacus_v3_chrX.gff")
dfgffref=as.data.frame(gff_ref)

dfgffref=as.data.frame(gff_ref)
dfgffref <- dfgffref %>% filter(str_detect(type, "gene")) %>% mutate(pstart=start-2000, pend=start+2000, pwidth=pend-pstart) %>% dplyr::select(-c(start, end, width)) #%>% rename(pstart="start", pend="end", pwidth="width")
# gene names to rownames 
rownames(dfgffref) <- dfgffref$ID
# make modified ref genome into dataframe
annoData2 <- makeGRangesFromDataFrame(dfgffref)
# annotate peaks
ol.anno2 <- annotatePeakInBatch(GRref, AnnotationData = annoData2, output="overlapping", maxgap=0)

library(Repitools)
library(tidyverse)
# replace genome start and end positions of peaks with distances to corresponding gene
prom_peaks <- annoGR2DF(ol.anno2)

prom_peaks <- prom_peaks %>% drop_na() %>% 
  mutate(pstart=end_position-start, pend=end_position-end) %>% 
  mutate(pstart=replace(pstart, which(pstart > 4000), 4000)) %>% 
  mutate(pend=replace(pend, which(pend < 0), 0))

# prepare empty dataframe for depth scores
peak_depth <- data.frame(matrix(ncol = 2, nrow = 0))
colnames(peak_depth) <- c("position", "depth")

# go through 1:2000 & count how many times it is >= start & <=end
 # add count as row to dataframe
for (i in 1:4000) {
  depth <- prom_peaks %>% count(i<=pstart & i>=pend)
  peak_depth <- peak_depth %>% add_row(position=i, depth=depth$n[2])
}


# switch position for accuracy in plot
actual_pos=2000:-1999
peak_depth <- peak_depth %>% mutate(position=actual_pos)
head(peak_depth)
tail(peak_depth)

length(actual_pos)



dfmeth <- read.table("/Users/corbinian/Documents/projects/ATAC_seq/data/methylation/Plainsman_nickname_unzip.bismark.cov")
colnames(dfmeth) <- c("chr", "start", "end", "fraction", "yes", "no")

dfmeth <- dfmeth %>% filter(str_detect(chr, "chr")) %>% mutate(chr=str_replace(chr, "chr_", "chr"), total=yes+no) %>% filter(fraction>=25, total>=5)
GRmeth <- makeGRangesFromDataFrame(dfmeth)
# annotate methylation
ol.anno3 <- annotatePeakInBatch(GRmeth, AnnotationData = annoData2, output="overlapping", maxgap=0)

library(Repitools)
library(tidyverse)
# replace genome start and end positions of peaks with distances to corresponding gene
prom_meth <- annoGR2DF(ol.anno3)

prom_meth <- prom_meth %>% drop_na() %>% 
  mutate(pstart=end_position-start, pend=end_position-end) %>% 
  mutate(pstart=replace(pstart, which(pstart > 4000), 4000)) %>% 
  mutate(pend=replace(pend, which(pend < 0), 0))

# prepare empty dataframe for depth scores
meth_depth <- data.frame(matrix(ncol = 2, nrow = 0))
colnames(meth_depth) <- c("position", "depth")

# go through 1:2000 & count how many times it is >= start & <=end
 # add count as row to dataframe
for (i in 1:4000) {
  depth2 <- prom_meth %>% count(i<=pstart & i>=pend)
  meth_depth <- meth_depth %>% add_row(position=i, depth=depth2$n[2])
}


# switch position for accuracy in plot
actual_pos=2000:-1999
meth_depth <- meth_depth %>% mutate(position=actual_pos)

meth_depth2 <- meth_depth %>% mutate(type="methylation")
peak_depth2 <- peak_depth %>% mutate(type="ACRs")
dfall_depth <- rbind(peak_depth2, meth_depth2)

#write_csv(dfall_depth, "~/Documents/projects/ATAC_seq/nature_comm_publication/data/Figure2A.csv")

# plot
#peak_depth %>% ggplot(aes(x=position, y=depth)) + geom_point()

scale_factor <- max(peak_depth$depth)/max(meth_depth$depth)
peak_depth %>% ggplot(aes(x=position, y=depth)) + geom_line(size=2, colour="lightcoral") +
  geom_line(data=meth_depth, aes(x=position, y=depth*scale_factor), colour="turquoise", alpha=0.75)+
  theme(axis.text.x = element_text(face="bold", size=35, angle=45, hjust=1),
        axis.title=element_text(size=45,face="bold"),
        axis.text.y = element_text(face="bold", size=35)) +
  scale_y_continuous(name = "OCR coverage",
                     sec.axis=sec_axis(~ . / scale_factor, name = "Methylation coverage")) +
  xlab("Distance to TSS (bp)") 



ggsave(filename = "~/Documents/projects/ATAC_seq/figures/ahypv3/meth_and_peak_depth_near_TSS_filtered.png", width=20, height=15) 
```
